Hola! Como te va?

Mi nombre es Facundo! Ya he tenido el agreago de haber revisado algunos de tus proyectos, hoy seré tu revisor nuevamente en este proyecto.

A continuación un poco sobre la modalidad de revisión que usaremos:

Cuando enccuentro un error por primera vez, simplemente lo señalaré, te dejaré encontrarlo y arreglarlo tú cuenta. Además, a lo largo del texto iré haciendo algunas observaciones sobre mejora en tu código y también haré comentarios sobre tus percepciones sobre el tema. Pero si aún no puedes realizar esta tarea, te daré una pista más precisa en la próxima iteración y también algunos ejemplos prácticos. Estaré abierto a comentarios y discusiones sobre el tema.

Encontrará mis comentarios a continuación: **no los mueva, modifique ni elimine**.

Puedes encontrar mis comentarios en cuadros verdes, amarillos o rojos como este:

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Exito. Todo se ha hecho de forma exitosa.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Observación. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Necesita arreglos. Este apartado necesita algunas correcciones. El trabajo no puede ser aceptado con comentarios rojos. 
</div>

Puedes responder utilizando esto:

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>
</div>

Estamos buscando para la empresa OilyGiant los mejores lugares donde abrir 200 pozos nuevos de petróleo.

 <div class="alert alert-block alert-success">
<b>Review General. (Iteración 3) </b> <a class="tocSkip"></a>

Felicitaciones, has implementado los cambios que te he indicado, bien hecho! Hemos alcanzado los objetivos del proyecto por lo que está en condiciones de ser aprobado!
    
Felicitaciones y éxitos en tu camino dentro del mundo de los datos!
    
Saludos!

<div class="alert alert-block alert-warning">
<b>Review General. (Iteración 2) </b> <a class="tocSkip"></a>

Hola nuevamente! :) un gusto! Veo que hemos intentado de corregir lo indicado y nos hemos acercado bastante, sin embargo hubo una leve confisión en donde aplicar el cambio por lo que obtuvimos resultados erroneos, una vez volvamos a modificar esto deberíamos poder ver mejores valores.
    
Espero nuevamente a esta corrección :) éxitos!

<div class="alert alert-block alert-success">
<b>Review General. (Iteración 1) </b> <a class="tocSkip"></a>


Hola, siempre me tomo este tiempo al inicio de tu proyecto para comentar mis apreciaciones generales de esta iteración de tu entrega. 
    

Me gusta comenzar dando la bienvenida al mundo de los datos a los estudiantes, te deseo lo mejor y espero que consigas lograr tus objetivos. Personalmente me gusta brindar el siguiente consejo, "Está bien equivocarse, es normal y es lo mejor que te puede pasar. Aprendemos de los errores y eso te hará mejor programador ya que podrás descubrir cosas a medida que avances y son estas cosas las que te darán esa experiencia para ser un gran Data Scientist"
    
Ahora si yendo a esta notebook.  Quería felicitarte porque has realizado de forma excelente el trabajo, has implementado de forma perfecta cada apartado y con un gran compromiso a la hora de mostrar los resultados parciales. Solo hemos tenido un pequeño detalle pero que estoy seguro que te demorará unos segundos corregir.

Espero con ansias a nuestra próxima iteración, exitos y saludos!

Paso 1. Descarga y preparación de los datos.
1.1 Importar los datasets utilizados, y las librerías correspondientes.

In [1]:
# Librerías
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression

In [2]:
# Leemos los archivos
zona_0 = pd.read_csv(
    '/datasets/geo_data_0.csv')
zona_1 = pd.read_csv(
    '/datasets/geo_data_1.csv')
zona_2 = pd.read_csv(
    '/datasets/geo_data_2.csv')

1.2 Observar los datos que contienen los datasets y los tipos de datos.

In [3]:
display(zona_0)
zona_0.info()
display(zona_0.describe())
display(zona_0.duplicated().sum())

id        f0        f1        f2     product
0      txEyH  0.705745 -0.497823  1.221170  105.280062
1      2acmU  1.334711 -0.340164  4.365080   73.037750
2      409Wp  1.022732  0.151990  1.419926   85.265647
3      iJLyR -0.032172  0.139033  2.978566  168.620776
4      Xdl7t  1.988431  0.155413  4.751769  154.036647
...      ...       ...       ...       ...         ...
99995  DLsed  0.971957  0.370953  6.075346  110.744026
99996  QKivN  1.392429 -0.382606  1.273912  122.346843
99997  3rnvd  1.029585  0.018787 -1.348308   64.375443
99998  7kl59  0.998163 -0.528582  1.583869   74.040764
99999  1CWhH  1.764754 -0.266417  5.722849  149.633246

[100000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

0

In [4]:
display(zona_1)
zona_1.info()
display(zona_1.describe())
display(zona_1.duplicated().sum())

id         f0         f1        f2     product
0      kBEdx -15.001348  -8.276000 -0.005876    3.179103
1      62mP7  14.272088  -3.475083  0.999183   26.953261
2      vyE1P   6.263187  -5.948386  5.001160  134.766305
3      KcrkZ -13.081196 -11.506057  4.999415  137.945408
4      AHL4O  12.702195  -8.147433  5.004363  134.766305
...      ...        ...        ...       ...         ...
99995  QywKC   9.535637  -6.878139  1.998296   53.906522
99996  ptvty -10.160631 -12.558096  5.005581  137.945408
99997  09gWa  -7.378891  -3.084104  4.998651  137.945408
99998  rqwUm   0.665714  -6.152593  1.000146   30.132364
99999  relB0  -3.426139  -7.794274 -0.003299    3.179103

[100000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

0

In [5]:
display(zona_2)
zona_2.info()
display(zona_2.describe())
display(zona_2.duplicated().sum())

id        f0        f1        f2     product
0      fwXo0 -1.146987  0.963328 -0.828965   27.758673
1      WJtFt  0.262778  0.269839 -2.530187   56.069697
2      ovLUW  0.194587  0.289035 -5.586433   62.871910
3      q6cA6  2.236060 -0.553760  0.930038  114.572842
4      WPMUX -0.515993  1.716266  5.899011  149.600746
...      ...       ...       ...       ...         ...
99995  4GxBu -1.777037  1.125220  6.263374  172.327046
99996  YKFjq -1.261523 -0.894828  2.524545  138.748846
99997  tKPY3 -1.199934 -2.957637  5.219411  157.080080
99998  nmxp2 -2.419896  2.417221 -5.548444   51.795253
99999  V9kWn -2.551421 -2.025625  6.090891  102.775767

[100000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

0

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Excelente carga de los datos manteniendolos por separados de las importaciones de librerías como así la implementación de métodos para profundizar en la comprensión de nuestros datos. Bien hecho!

En los dataframes de las 3 regiones (zona_1, zona_2 y zona_3) vemos que no hay datos nulos ni duplicados. A parte de la columna con los ID's de las pozos, los demás datos poseen datos de tipo float64. Con las columnas f0, f1 y f2 vemos que los datos mínimos son negativos en todos los Dataframes. Algunos promedios también son negativos. Todos los máximos son números positivos.

1.3 Preprocesamiento de los datos.

Separaremos en datasets a parte los datos que no parecen ser necesarios para nuestro modelo, los datos de las columnas id. En los datasets originales eliminaremos dichas columnas.

In [6]:
id_0 = zona_0[['id']]
id_1 = zona_1[['id']]
id_2 = zona_2[['id']]

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Excelente decisión al eliminar aquella feature que no aporta valor a nuestro modelo!

Estandarización de los datos.

In [7]:
scaler = StandardScaler()

# Estandarizamos los dataframes.
zona_1_stan = scaler.fit_transform(zona_1.drop(columns=['id']))
zona_2_stan = scaler.fit_transform(zona_2.drop(columns=['id']))
zona_0_stan = scaler.fit_transform(zona_0.drop(columns=['id']))

# Guardar los parámetros del scaler para revertir la estandarización más adelante
scaler_mean = scaler.mean_
scaler_scale = scaler.scale_

# Convertir a DataFrame y agregar la columna 'product'
z_0 = pd.DataFrame(
    zona_0_stan, columns=zona_0.columns[1:], index=zona_0.index)
z_1 = pd.DataFrame(
    zona_1_stan, columns=zona_1.columns[1:], index=zona_1.index)
z_2 = pd.DataFrame(
    zona_2_stan, columns=zona_2.columns[1:], index=zona_2.index)

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Y tal como siempre se recomiendo hemos escalada nuestros datos mediante el uso de StandarScaler, bien hecho!

Paso 2. Entrenar y poner a prueba el modelo de regresión lineal para cada región.

Hacemos una función con la cual dividimos los datos en conjunto de entrenamiento y validación (en una relación 75:25), entrenamos el modelo de regresión lineal para cada región.

In [8]:
def train_and_evaluate(data, scaler_mean, scaler_scale):
    # Dividir los datos
    features = data[['f0', 'f1', 'f2']]
    target = data['product']
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target, test_size=0.25, random_state=12345)

    # Entrenar el modelo
    model = LinearRegression()
    model.fit(features_train, target_train)

    # Hacer predicciones
    target_pred = model.predict(features_valid)

    # Calcular el RMSE
    rmse = np.sqrt(mean_squared_error(target_valid, target_pred))

    # Revertir la estandarización de las predicciones y los valores reales
    target_pred_orig = target_pred * scaler_scale[-1] + scaler_mean[-1]
    target_valid_orig = target_valid * scaler_scale[-1] + scaler_mean[-1]
    # Mantenemos los índices originales y pasamos las predicciones a una Serie de Pandas
    target_pred_orig = pd.Series(target_pred_orig, index=features_valid.index)
    target_valid_orig = pd.Series(
        target_valid_orig, index=features_valid.index)

    return target_pred, target_valid, rmse, target_pred_orig, target_valid_orig

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Perfecta creación de las funciones, esto nos permitirá ser más ordenados con nuestro código, perfecta división de los datos y una  excelente declaración del modelo junto al entrenamiento y predicción! A la vez excelente elección de métrica!

Evaluamos el modelo

Analizamos la raíz del error cuadrático medio y el promedio de reservas predicho.

In [9]:
# Evaluar cada región
target_pred_0, target_valid_0, rmse_0, target_pred_orig_0, target_valid_orig_0 = train_and_evaluate(
    z_0, scaler_mean, scaler_scale)
target_pred_1, target_valid_1, rmse_1, target_pred_orig_1, target_valid_orig_1 = train_and_evaluate(
    z_1, scaler_mean, scaler_scale)
target_pred_2, target_valid_2, rmse_2, target_pred_orig_2, target_valid_orig_2 = train_and_evaluate(
    z_2, scaler_mean, scaler_scale)

display(f"RMSE en la Región 0: {rmse_0}")
display(f"RMSE en la Región 1: {rmse_1}")
display(f"RMSE en la Región 2: {rmse_2}")

display(
    f"Reservas promedio predichas originales en la Región 0: {target_pred_orig_0.mean()}")
display(
    f"Reservas promedio predichas originales en la Región 1: {target_pred_orig_1.mean()}")
display(
    f"Reservas promedio predichas originales en la Región 2: {target_pred_orig_2.mean()}")

display(
    f"Reservas promedio reales originales en la Región 0: {target_valid_orig_0.mean()}")
display(
    f"Reservas promedio reales originales en la Región 1: {target_valid_orig_1.mean()}")
display(
    f"Reservas promedio reales originales en la Región 2: {target_valid_orig_2.mean()}")

'RMSE en la Región 0: 0.8485147934825391'

'RMSE en la Región 1: 0.01943878486290051'

'RMSE en la Región 2: 0.894524681335185'

'Reservas promedio predichas originales en la Región 0: 92.59256778438035'

'Reservas promedio predichas originales en la Región 1: 92.40702284439296'

'Reservas promedio predichas originales en la Región 2: 92.46540623335103'

'Reservas promedio reales originales en la Región 0: 92.07859674082927'

'Reservas promedio reales originales en la Región 1: 92.4018069687064'

'Reservas promedio reales originales en la Región 2: 92.38542600187984'

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Predicciones excelentes, llegamos a los resultados esperados!

El modelo se ajusta mejor a los datos de la region 1, por lo que es más preciso en las predicciones. El RMSE está en las mismas unidades que la variable objetivo, en este este caso miles de barriles", aunque estas unidades fueron estandarizadas.

Hay que considerar que nuestros datos tienen un rango de 8 unidades (-4 a 4, aproximadamente). En promedio, las predicciones del modelo se desvían en aproximadamente 0.02 unidades de los valores reales en la region 1, 0.85 en la region 0 y 0.9 en la región 2. Esto podemos compararlo con los datos originales reales.

Estos resultados sugieren que el modelo está realizando predicciones con un error promedio pequeño en comparación con la magnitud total de los datos. El modelo se ajusta bien a los datos en el contexto de su rango.

De acuerdo a las reservas predichas usando las unidades originales de miles de barriles, encontramos que cada una de las tres regiones en promedio tienen alrededor de 92 mil barriles de producción.

Paso 3. Preparación para el cálculo de ganancias

Contamos con un presupuesto de 100 millones de dólares, el cual será usado para desarrollar la explotación de 200 pozos petroleros en total. Un barril de materias primas genera 4.5 USD de ingresos. El ingreso de una unidad de producto es de 4500 dólares (el volumen de reservas está expresado en miles de barriles). De media un pozo debe producir al menos un valor de 500,000 dólares en unidades para evitar pérdidas (esto es equivalente a 111.1 unidades). 

In [10]:
BUDGET = 100e6  # Presupuesto
WELL_COST = BUDGET / 200  # Costo estipulado para cada pozo
REVENUE_PER_UNIT = 4500

threshold = WELL_COST / REVENUE_PER_UNIT
display(f"Umbral de reservas: {threshold}")

'Umbral de reservas: 111.11111111111111'

Para calcular las ganancias deberemos multuplicar los ingresos por la unidad de producto (REVENUE_PER_UNIT) por las reservas totales luego de sumar las reservas de los 200 pozos de acuerdo a las predicciones más altas.

Paso 4. Escribir una función para calcular la ganancia de un conjunto de pozos de petróleo seleccionados. Modelar las predicciones.

In [11]:
def calculate_profit(predictions, target):
    # Seleccionar los 200 pozos con las predicciones más altas
    selected = predictions.nlargest(200)

    # Asegurarse de que los índices se alinean correctamente
    selected_index = selected.index.intersection(target.index)

    # Calcular el volumen total de reservas
    total_reserves = target.loc[selected_index].sum()

    # Calcular las ganancias
    revenue = total_reserves * REVENUE_PER_UNIT
    profit = revenue - BUDGET

    return profit, selected

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Perfecto! Tal como se debía hemos tomado los top 200 pozos en base a las predicciones eligiendo los valores reales y de forma correcta hemos calculado la ganancia al multiplicar por la ganancia por unidad y al restar el presupuesto. Sigamos!

In [12]:
# Calcular la ganancia para cada región
profit_0, selected_predictions_0 = calculate_profit(
    target_pred_orig_0, target_valid_orig_0)
profit_1, selected_predictions_1 = calculate_profit(
    target_pred_orig_1, target_valid_orig_1)
profit_2, selected_predictions_2 = calculate_profit(
    target_pred_orig_2, target_valid_orig_2)

# Diccionario de ganancias
profits = {
    "Región 0": profit_0,
    "Región 1": profit_1,
    "Región 2": profit_2
}

# Encontrar la región con la ganancia máxima
max_region = max(profits, key=profits.get)
max_profit = profits[max_region]

# Mostrar todas las ganancias y las regiones
for region, profit in profits.items():
    display(f"Ganancia en la {region}: {profit} dólares")

# Mostrar la region con la mayor ganancia
display(f"La mayor ganancia es en la {max_region} con: {max_profit} dólares")

'Ganancia en la Región 0: 33208260.43139851 dólares'

'Ganancia en la Región 1: 43216519.919496775 dólares'

'Ganancia en la Región 2: 24424699.341968343 dólares'

'La mayor ganancia es en la Región 1 con: 43216519.919496775 dólares'

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Obtenciones perfectas

Observamos que la mayor ganancia prevista está en los 200 pozos de la región 1, con una estimación de 43216520 dólares en ganancias aproximadamente. Proponemos los 200 pozos de esta región para su desarrollo.

Paso 5: Análisis de riesgos y ganancias

In [13]:
def bootstrap_profit_risk(selected_predictions, target, n_samples=1000):
    state = np.random.RandomState(12345)
    profits = []

    for _ in range(n_samples):
        # Tomar una muestra aleatoria con reemplazo
        target_subsample = target.sample(
            n=500, replace=True, random_state=state)
        predictions_subsample = selected_predictions.reindex(
            target_subsample.index)

        # Calcular la ganancia usando las predicciones seleccionadas
        profit, _ = calculate_profit(predictions_subsample, target_subsample)
        profits.append(profit)

    # Convertir la lista de ganancias en un Series de pandas
    profits_series = pd.Series(profits)

    # Calcular el intervalo de confianza del 95%
    lower = profits_series.quantile(0.025)
    upper = profits_series.quantile(0.975)

    # Calcular el beneficio promedio
    mean_profits_series = profits_series.mean()

    # Calcular el riesgo de pérdida como porcentaje
    risk_loss_percentage = (profits_series < 0).mean() * 100

    return mean_profits_series, lower, upper, risk_loss_percentage

In [16]:
# Calcular la ganancia y el riesgo para la región 1 usando bootstrapping
mean_profit_0, lower_bound_0, upper_bound_0, loss_percentage_0 = bootstrap_profit_risk(
    target_pred_orig_0, target_valid_orig_0)

mean_profit_1, lower_bound_1, upper_bound_1, loss_percentage_1 = bootstrap_profit_risk(
    target_pred_orig_1, target_valid_orig_1)

mean_profit_2, lower_bound_2, upper_bound_2, loss_percentage_2 = bootstrap_profit_risk(
    target_pred_orig_2, target_valid_orig_2)

display(
    f"Región 0 - Beneficio promedio: {mean_profit_0:.2f} Intervalo de confianza del 95%: ({lower_bound_0:.2f}, {upper_bound_0:.2f}) Probabilidad de pérdida: {loss_percentage_0:.2f}%")
display(
    f"Región 1 - Beneficio promedio: {mean_profit_1:.2f} Intervalo de confianza del 95%: ({lower_bound_1:.2f}, {upper_bound_1:.2f}) Probabilidad de pérdida: {loss_percentage_1:.2f}%")
display(
    f"Región 2 - Beneficio promedio: {mean_profit_2:.2f} Intervalo de confianza del 95%: ({lower_bound_2:.2f}, {upper_bound_2:.2f}) Probabilidad de pérdida: {loss_percentage_2:.2f}%")

'Región 0 - Beneficio promedio: 3965603.33 Intervalo de confianza del 95%: (-1112155.46, 9097669.42) Probabilidad de pérdida: 6.90%'

'Región 1 - Beneficio promedio: 24334963.62 Intervalo de confianza del 95%: (20262011.83, 28151743.82) Probabilidad de pérdida: 0.00%'

'Región 2 - Beneficio promedio: 1606133.62 Intervalo de confianza del 95%: (-4016116.95, 7006194.71) Probabilidad de pérdida: 26.80%'

<div class="alert alert-block alert-warning">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Aquí lo hemos hecho muy bien, la estructura es perfecta sin embargo nos estamos cconfunidiendo en un pequeño detalle, nosotors le estamos pasando unicamente los 200 pozos elegidos por ser top prediction cuando en realidad aquí lo que debemos hacer es pasar todos los valores reales de la región, el objetivo del bootstrapping es asegurar la consistencia en las ganancias por lo cual debemos tomar todos los datos de la región y sobre cada iteración tomar una submuestra pequeña.
    
Nosotros actualmente tomamos 200 valores y tomamos una submuestra de 200 valores por lo cual en cada iteración tendremos el mismo resultado, tratemos de corregir esto, tomemos todos los valores de la región y aumenta el sample que tomamos a 500 valores.

<div class="alert alert-block alert-info">
<b>¡Hola Facundo! Muchas gracias por tu respuesta. Tuve muchas dudas de como resolver esta parte del proyecto. En las instrucciones del proyecto se mencionaba que usaramos las predicciones almacenadas en el paso 4.2, por lo que me fui con la finta de que tenían que ser los 200 valores... se me hizo raro, pero lo que mencionas tiene más sentido. También sigo teniendo dudas al respecto, debido a que los resultados me aparecen negativos, con probabilidades de pérdidas del 100%. ¿Estoy malentendiedo otra cosa 

</b> <a class="tocSkip"></a>
</div>


<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Hola nuevamente! :) entiendo tu consulta, esto se debe a una leve confusión en lo indicado en la iteración anterior, los 500 valores se refieren a la muestra, nosotros estamos modificando la cantidad de iteraciones y aun seguimos tomando muestras de 200 casos, lo correcto sería iterar 1000 veces y tomar 500 valores por cada iteración, es decir:
    
    target.sample(
            n=500 ...
    
una vez modifiquemos esto veremos los resultados correctos!
    
    Ahora si excelente, felicitaciones!

Conclusión

In [15]:
if loss_percentage_0 < 0.025 and mean_profit_0 > mean_profit_1 and mean_profit_0 > mean_profit_2:
    selected_region = "Región 0"
elif loss_percentage_1 < 0.025 and mean_profit_1 > mean_profit_0 and mean_profit_1 > mean_profit_2:
    selected_region = "Región 1"
elif loss_percentage_2 < 0.025 and mean_profit_2 > mean_profit_0 and mean_profit_2 > mean_profit_1:
    selected_region = "Región 2"
else:
    selected_region = "Ninguna región cumple con los criterios de riesgo."

display(
    f"Región seleccionada para el desarrollo de pozos petrolíferos: {selected_region}")

'Región seleccionada para el desarrollo de pozos petrolíferos: Región 1'

Concluimos que el desarrollo de los pozos petrolíferos debe ser en la región 1, donde encontramos el menor riesgo de pérdidas y con un beneficio estimado de 24334963.62 dólares. Dichos resultados sobre la región donde explotar los pozos concuerdan con los obtenidos sobre el sitio con las mayores ganancias.
